File: wordBreak.cpp
#include <bits/stdc++.h>
using namespace std;
const int ALPHA = 26;

struct TrieNode
{
	struct TrieNode *child[ALPHA];
	int isEndOfWord;
};

struct TrieNode *getNode(void)
{
	struct TrieNode *pNode = new TrieNode();
	pNode->isEndOfWord = false;
	for (int i = 0; i < ALPHA; i++)
	{
		pNode->child[i] = NULL;
	}
	return pNode;
}

void insert(struct TrieNode *root, string key)
{
	struct TrieNode *pCrawl = root;
	for (int i = 0; i < key.length(); i++)
	{
		int index = key[i] - 'a';
		if (!pCrawl->child[index])
			pCrawl->child[index] = getNode();
		pCrawl = pCrawl->child[index];
	}
	pCrawl->isEndOfWord = true;
}

bool search(struct TrieNode *root, string str)
{
	struct TrieNode *pCrawl = root;
	for (int i = 0; i < str.length(); i++)
	{
		int index = str[i] - 'a';
		if (!pCrawl->child[index])
			return false;
		pCrawl = pCrawl->child[index];
	}
	return (pCrawl != NULL && pCrawl->isEndOfWord);
}

bool wordbreak(string str, struct TrieNode *root)
{
	int len = str.size();
	if (len == 0)
		return true;
	for (int i = 1; i <= len; i++)
	{
		if (search(root, str.substr(0, i)) && wordbreak(str.substr(i, len - i), root))
			return true;
	}
	return false;
}

int main()
{
	string dictionary[] = {"mobile", "samsung", "sam", "sung", "ma\n", "mango", "icecream", "and", "go", "i", "like", "ice", "cream"};
	int n = sizeof(dictionary) / sizeof(dictionary[0]);
	struct TrieNode *root = getNode();
	for (int i = 0; i < n; i++)
		insert(root, dictionary[i]);
	wordbreak("ilikesamsung", root) ? cout << "Yes\n" : cout << " No\n";
}

File: wordBreak.py
class TrieNode:
    def __init__(self):
        self.children = [None] * 26
        self.isEndOfWord = False


class Trie:
    def __init__(self):
        self.root = self.getNode()

    def getNode(self):
        return TrieNode()

    def insert(self, key):
        pCrawl = self.root
        n = len(key)
        for i in range(n):
            index = ord(key[i]) - ord('a')
            if not pCrawl.children[index]:
                pCrawl.children[index] = self.getNode()
            pCrawl = pCrawl.children[index]
        pCrawl.isEndOfWord = True

    def search(self, item, root):
        n = len(item)
        pCrawl = root
        for i in range(n):
            index = ord(item[i]) - ord('a')
            if not pCrawl.children[index]:
                return False
            pCrawl = pCrawl.children[index]
        return pCrawl != None and pCrawl.isEndOfWord

    def checkWordBreak(self, root, strr):
        n = len(strr)
        if n == 0:
            return True
        for i in range(1, n+1):
            if(self.search(strr[:i], root) and self.checkWordBreak(root, strr[i:])):
                return True
        return False

    def wordBreak(self, s):
        pNode = self.root
        if self.checkWordBreak(pNode, s):
            return True
        else:
            return False


def main():
    keys = ["the", "quick", "fox", "brown"]
    s = "thequickthe"
    t = Trie()
    for key in keys:
        t.insert(key)
    # print(output[t.search("the")])
    # print(output[t.search("these")])
    if t.wordBreak(s):
        print("Yes")
    else:
        print("No")


if __name__ == '__main__':
    main()
